engines/dfc_provider/app/services/affiliate_sales_query.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+class AffiliateSalesQuery
+  class << self
+    def data(enterprises)
+      Spree::LineItem
+        .joins(tables)
+        .where(
+          spree_orders: { state: "complete", distributor_id: enterprises },
+        )
+        .group(key_fields)
+        .pluck(fields)
+    end
+
+    # Create a hash with labels for an array of data points:
+    #
+    #   { product_name: "Apple", ... }
+    def label_row(row)
+      labels.zip(row).to_h
```

Comment:
Ah, I didn't know `select_all`. I'll keep that in mind for another time. If it could do it in batches then I would probably prefer that. But my version occupies the least memory for grabbing all data at once, I think, because it's only keeping the values in an array and the keys are only added while processing one row, as you acknowledged above.